Fix dynamic linking issues with prebuilt pip packages

dbort opened the pull request:
We can now prebuild pip packages, but we're hitting issues when importing pybindings and torch at runtime.

More context in [T185860350](https://www.internalfb.com/intern/tasks/?t=185860350)

## Test Plan
- Ensure that the PR is against `release/0.2`
- PR must touch one of the files in the `pull_request.paths` entry for a given build-wheels workflow: https://github.com/pytorch/executorch/blob/90d0c1a1fb94503e2f4d2b938d26ceae1d60492b/.github/workflows/build-wheels-m1.yml#L5-L8
  - This is necessary to run the build-wheel jobs in the `test` channel (using the pytorch release pip package) instead of the `nightly` channel
- Add the label `ciflow/binaries/all` to the PR so that it builds for all versions of python and not just 3.8 (which doesn't work for executorch)
- After the artifacts are built, follow the instructions in the comment below: https://github.com/pytorch/executorch/pull/3049#issuecomment-2065488712
- Also, this PR adds a smoke test that will test the wheels during the build job

dbort committed:
Build pybindings with -D_GLIBCXX_USE_CXX11_ABI=0 to match libtorch.so

libtorch.so builds with the old glibc ABI, so we need to as well,
for any source files that include torch headers.

dbort commented:
Building wheels:
- macos: https://github.com/pytorch/executorch/actions/runs/8697807901
  - Triggered from https://github.com/pytorch/executorch/actions/workflows/build-wheels-m1.yml- linux: https://github.com/pytorch/executorch/actions/runs/8697825002
  - Triggered from https://github.com/pytorch/executorch/actions/workflows/build-wheels-linux.yml

dbort commented:
To test out a prebuilt wheel on Linux:
- \[Optional\] Open a new terminal: the script will create a new conda environment called `executorch-temp`, so using a new terminal will help ensure that you don't accidentally continue working in that env afterwards.
- Ensure that you have conda installed
- Click this link to download a zipped wheel file: [pytorch_executorch__3.10_cpu_x86_64.zip](https://github.com/pytorch/executorch/actions/runs/8745358205/artifacts/1427679533)
  - If you're dowloading this on your mac you'll need to `scp` the .zip file to your devserver, like `scp ~/Downloads/pytorch_executorch__3.10_cpu_x86_64.zip MYDEVERSVER:/tmp/`
- Download the test script: `curl -o /tmp/test_wheel_install.sh https://raw.githubusercontent.com/pytorch/executorch/sandbox/dbort/pip-linking/build/test_wheel_install.sh`
- `cd` to the root of a relatively recent `executorch` clone. Exact version doesn't matter.
- Run the script against the zip file you downloaded: `bash /tmp/test_wheel_install.sh /path/to/pytorch_executorch__3.10_cpu_x86_64.zip | tee /tmp/wheel-install-logs`

If you see `PASS` at the end, it worked! If not, please create a paste of `/tmp/wheel-install-logs` and let me know.

```
+ python /tmp/test-wheel-install-lWkozN8xUl/test_xnnpack_e2e.py
I 00:00:00.003848 executorch:program.cpp:129] InternalConsistency verification requested but not available
E 00:00:00.023344 executorch:method.cpp:936] Overriding output data pointer allocated by memory plan is not allowed.
I 00:00:00.023371 executorch:pybindings.cpp:196] Cannot set_output_data_ptr(): this likely means the outputs were MemoryPlanned inspect the error code to know for sure, but likely this is not an issue. 0x2
PASS <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
+ rm -rf /tmp/test-wheel-install-lWkozN8xUl
```

dbort committed:
Set the RPATH of _portable_lib.so so it can find libtorch

pip wheels will need to be able to find the torch libraries. On Linux,
the .so has non-absolute dependencies on libs like "libtorch.so" without
paths; as long as we `import torch` first, those dependencies will work.

But Apple dylibs do not support non-absolute dependencies, so we need
to tell the loader where to look for its libraries. The LC_LOAD_DYLIB
entries for the torch libraries will look like "@rpath/libtorch.dylib",
so we can add an LC_RPATH entry to look in a directory relative to the
installed location of our _portable_lib.so file.

To see these LC_* values, run `otool -l _portable_lib*.so`.

dbort committed:
Disable wheel delocation on macos

The executorch build system will ensure that .dylib/.so files have
LC_LOAD_DYLIB and LC_RPATH entries that will work when they're
installed.

Delocating (i.e., making copies of the .dylibs that ET's libs depend on)
will break any libs that depend on the torch libraries if users ever
import both `torch` and the executorch library. Both import paths must
load exactly the same file, not just a copy of it.

kirklandsign commented:
PASS
Apple M1 Pro
Darwin Kernel Version 23.4.0

larryliu0820 commented:
I ran into some issues on my Mac m1: https://gist.github.com/larryliu0820/7f95f711c2a40835c3e39680a7a6e44d

dbort committed:
Implemement smoke_test.py for pip wheel jobs

This script is run by CI after building the executorch wheel. Before
running this, the job will install the matching torch package as well as
the newly-built executorch package and its dependencies.

For now we test the export of a simple model, and try executing it using
the runtime pybindings.

Test Plan:
```
./install_requirements.sh
python build/packaging/smoke_test.py
```

dbort commented:
Added a smoke test script to the wheel-build job: succeeding for the expected-compatible python versions, like [3.10 for linux](https://github.com/pytorch/executorch/actions/runs/8789317343/job/24118786907?pr=3049#step:16:1226) and [3.10 for mac](https://github.com/pytorch/executorch/actions/runs/8789317334/job/24118786398?pr=3049#step:15:1197)

larryliu0820 reviewed (approved):
Thanks for adding the test

dbort commented:
Successfully ran the test script under WSL running Ubuntu 20.04.4 LTS on Windows 10 22H2, AMD Ryzen 7 2700 Eight-Core Processor 3.20 GHz, 32GB RAM

guangy10 commented:
@dbort We expect the prebuilt works for 3.10 only right?

dbort commented:
> @dbort We expect the prebuilt works for 3.10 only right?

That's the main version that we need to work. Currently there's no way to disable the python versions that we don't expect to work (like 3.8 and 3.12).
